checker/src/org/checkerframework/checker/guieffect/GuiEffectVisitor.java

```diff
@@ -158,16 +227,75 @@ public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
             // Instantiate type-polymorphic effects
             if (srcType.hasAnnotation(AlwaysSafe.class)) {
                 targetEffect = new Effect(SafeEffect.class);
+                polyUIArgumentsAreUI = false;
             } else if (srcType.hasAnnotation(UI.class)) {
                 targetEffect = new Effect(UIEffect.class);
             }
             // Poly substitution would be a noop.
         }
 
-        Effect callerEffect = atypeFactory.getDeclaredEffect(callerElt);
-        // Field initializers inside anonymous inner classes show up with a null current-method ---
-        // the traversal goes straight from the class to the initializer.
-        assert (currentMethods.peek() == null || callerEffect.equals(effStack.peek()));
+        // Look for lambda arguments and infer them to be @UIEffect for this call
```

Comment:
I'm wary of mixing inference and checking in the same pass (lines 237-280) since we're not doing full effect inference.  As I've mentioned elsewhere, the inference work should live in the TypeFactory and TreeAnnotator, updating the tree itself.  The separation makes it easier to debug later.